### PR TITLE
Learnable Per-Domain Cp Scale: self-calibrating physics hint

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Panel Cp feature uses a fixed scale=0.1 found empirically by PR #2319 (after trying 1.0 and 0.0). A **learnable scale parameter per domain** (tandem vs single-foil), initialized to 0.1, lets the model self-calibrate the Cp signal strength during training. The tandem domain may benefit from a different effective scale than single-foil.

This removes a hand-tuned hyperparameter and may find a better operating point. Low risk, low complexity (~20 LoC).

### Changes to `cfd_tandemfoil/train.py`

1. **Add flag:**
   ```python
   parser.add_argument('--cp_panel_learnable_scale', action='store_true')
   ```

2. **Replace fixed cp_panel_scale with learnable parameter:**
   ```python
   if args.cp_panel_learnable_scale:
       # Learnable log-scale, initialized so sigmoid(x)*0.2 ≈ 0.1
       # sigmoid(-0.41) ≈ 0.4, 0.4 * 0.2 = 0.08 ≈ 0.1
       self.cp_log_scale_tandem = nn.Parameter(torch.tensor(0.0))  
       # Effective scale = sigmoid(param) * 0.2, range (0, 0.2)
       # At init: sigmoid(0) * 0.2 = 0.1 (matches current baseline)
   ```

3. **In feature construction:**
   ```python
   if args.cp_panel_learnable_scale:
       eff_scale = torch.sigmoid(self.cp_log_scale_tandem) * 0.2
       cp_features = cp_features * eff_scale
   else:
       cp_features = cp_features * args.cp_panel_scale
   ```

4. **Log learned scale to W&B:**
   ```python
   wandb.log({'train/cp_learned_scale': (torch.sigmoid(self.cp_log_scale_tandem) * 0.2).item()})
   ```

5. **Run 2 seeds:**
   ```bash
   CUDA_VISIBLE_DEVICES=0 python train.py \
     --agent nezuko --wandb_name "nezuko/learnable-cp-s42" --wandb_group learnable-cp-scale \
     --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
     --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
     --pressure_first --pressure_deep --residual_prediction --surface_refine \
     --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
     --cp_panel --cp_panel_tandem_only --cp_panel_learnable_scale

   CUDA_VISIBLE_DEVICES=1 python train.py \
     --agent nezuko --wandb_name "nezuko/learnable-cp-s73" --wandb_group learnable-cp-scale \
     --seed 73 [same flags]
   ```

   Note: do NOT pass --cp_panel_scale when using --cp_panel_learnable_scale (the learned param replaces it).

## Baseline
| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| **p_in** | **11.709** | < 11.71 |
| **p_oodc** | **7.544** | < 7.54 |
| **p_tan** | **27.402** | < 27.40 |
| p_re | 6.481 | < 6.48 |

W&B baseline: h6fqcry4 (s42), cuhoscp9 (s73)
Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Learnable Per-Domain Cp Scale

| Metric | s42 | s73 | 2-seed avg | Baseline | Delta |
|--------|-----|-----|------------|----------|-------|
| **p_in** | 12.498 | 12.316 | **12.407** | 11.709 | +5.97% ❌ |
| **p_oodc** | 11.940 | 9.436 | **10.688** | 7.544 | +41.7% ❌ |
| **p_tan** | 38.400 | 34.376 | **36.388** | 27.402 | +32.8% ❌ |
| **p_re** | 8.092 | 7.073 | **7.583** | 6.481 | +17.0% ❌ |

**All metrics significantly worse than baseline. Experiment did not beat any target.**

### Learned Cp Scale
- s42: converged to **0.0929** (from init 0.1)
- s73: converged to **0.0865** (from init 0.1)
- Average: **0.0897** — very close to the fixed 0.1, so the model didn't find a meaningfully different optimal scale.

### W&B Runs
- s42: `v9q38xst` (best epoch 160)
- s73: `5g5sr9cg` (best epoch 161)
- Group: `learnable-cp-scale`

### Command Used
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent nezuko --wandb_name "nezuko/learnable-cp-s42" --wandb_group learnable-cp-scale \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_learnable_scale
```

### What Happened

The hypothesis that a learnable Cp scale would self-calibrate and find a better operating point than the fixed 0.1 did not hold. Key observations:

1. **The learned scale converged near the fixed value** (~0.09 vs 0.1), suggesting 0.1 was already near optimal. The model didn't discover a substantially different preferred scale.

2. **Despite converging to nearly the same scale, performance degraded massively** — especially on tandem (+33%) and OOD-condition (+42%) splits, exactly where the Cp panel feature is active (tandem-only). This strongly suggests the issue isn't the scale's final value, but the optimization dynamics of learning it.

3. **Likely cause: interference with PCGrad 3-way gradient surgery.** The learnable scale parameter creates a shared bottleneck through which gradients from all three tasks (surface, volume, coarse) flow. PCGrad projects conflicting gradients, but this shared parameter makes conflicts worse — each task tries to adjust the Cp scale differently, causing unstable optimization.

4. **The sigmoid(x)×0.2 parameterization may have created a near-flat gradient landscape.** At the init point (sigmoid(0)=0.5), the sigmoid derivative is maximal, but the 0.2 multiplier compresses the effective gradient. The parameter barely moved from its initialization, suggesting weak gradient signal.

### Suggested Follow-ups

- **Don't make the Cp scale learnable.** The fixed 0.1 is already well-calibrated, and introducing a learnable parameter hurts optimization dynamics.
- If the scale must be tuned, a simpler **grid search over fixed values** (e.g. 0.05, 0.1, 0.15, 0.2) would be safer and avoid PCGrad interference.
- The large degradation from such a small change (~20 LoC, ~1 parameter) highlights how sensitive this model is to optimization dynamics with PCGrad. Future experiments should be cautious about adding learnable parameters that create shared gradient paths.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2353 because ADVISOR: Closing — catastrophic regression across all metrics (+6-42%). The learned scale converged near 0.09 (close to the fixed 0.1 we already use), confirming our manual choice was already near-optimal. More critically, making the scale learnable introduced destructive optimization dynamics with PCGrad — the gradients from Cp loss fought against the main surface loss during early training when the scale was far from optimal. Fixed scaling at 0.1 remains the right approach. Thank you for the thorough investigation, nezuko..
